### DISCUSSION TITLE
Can't use pip install after cloned

### DISCUSSION THREAD

**Discussion by paemaker** [nasa/fprime#1357] [Q&A] [answered]
I tried to use `pip install --upgrade fprime-tools fprime-gds` but it didn't work out. It's showing

>  error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
>   ERROR: Failed building wheel for lxml
>   Running setup.py clean for lxml
> Failed to build lxml
> Installing collected packages: lxml, Jinja2, itsdangerous, idna, colorama, Cheetah3, charset-normalizer, chardet, attrs, atomicwrites, requests, pytest, click, binaryornot, arrow, jinja2-time, flask, flask-restful, cookiecutter, fprime-tools, fprime-gds
>   Running setup.py install for lxml ... error
>   error: subprocess-exited-with-error
> 
>   × Running setup.py install for lxml did not run successfully.
>   │ exit code: 1
>   ╰─> [76 lines of output]
>       Building lxml version 4.6.3.
>       Building without Cython.
>       Building against pre-built libxml2 andl libxslt libraries
>       running install
>       D:\Programs\Python\lib\site-packages\setuptools\command\install.py:34: SetuptoolsDeprecationWarning: setup.py install is deprecated. Use build and pip and other standards-based tools.
>         warnings.warn(
>       running build
>       running build_py
>       creating build
>       creating build\lib.win-amd64-3.10
>       creating build\lib.win-amd64-3.10\lxml
>       copying src\lxml\builder.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\cssselect.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\doctestcompare.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\ElementInclude.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\pyclasslookup.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\sax.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\usedoctest.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\_elementpath.py -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\__init__.py -> build\lib.win-amd64-3.10\lxml
>       creating build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\__init__.py -> build\lib.win-amd64-3.10\lxml\includes
>       creating build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\builder.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\clean.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\defs.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\diff.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\ElementSoup.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\formfill.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\html5parser.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\soupparser.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\usedoctest.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\_diffcommand.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\_html5builder.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\_setmixin.py -> build\lib.win-amd64-3.10\lxml\html
>       copying src\lxml\html\__init__.py -> build\lib.win-amd64-3.10\lxml\html
>       creating build\lib.win-amd64-3.10\lxml\isoschematron
>       copying src\lxml\isoschematron\__init__.py -> build\lib.win-amd64-3.10\lxml\isoschematron
>       copying src\lxml\etree.h -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\etree_api.h -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\lxml.etree.h -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\lxml.etree_api.h -> build\lib.win-amd64-3.10\lxml
>       copying src\lxml\includes\c14n.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\config.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\dtdvalid.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\etreepublic.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\htmlparser.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\relaxng.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\schematron.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\tree.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\uri.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\xinclude.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\xmlerror.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\xmlparser.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\xmlschema.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\xpath.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\xslt.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\__init__.pxd -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\etree_defs.h -> build\lib.win-amd64-3.10\lxml\includes
>       copying src\lxml\includes\lxml-version.h -> build\lib.win-amd64-3.10\lxml\includes
>       creating build\lib.win-amd64-3.10\lxml\isoschematron\resources
>       creating build\lib.win-amd64-3.10\lxml\isoschematron\resources\rng
>       copying src\lxml\isoschematron\resources\rng\iso-schematron.rng -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\rng
>       creating build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl
>       copying src\lxml\isoschematron\resources\xsl\RNG2Schtrn.xsl -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl
>       copying src\lxml\isoschematron\resources\xsl\XSD2Schtrn.xsl -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl
>       creating build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
>       copying src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_abstract_expand.xsl -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
>       copying src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_dsdl_include.xsl -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
>       copying src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_schematron_message.xsl -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
>       copying src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_schematron_skeleton_for_xslt1.xsl -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
>       copying src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\iso_svrl_for_xslt1.xsl -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
>       copying src\lxml\isoschematron\resources\xsl\iso-schematron-xslt1\readme.txt -> build\lib.win-amd64-3.10\lxml\isoschematron\resources\xsl\iso-schematron-xslt1
>       running build_ext
>       building 'lxml.etree' extension
>       error: Microsoft Visual C++ 14.0 or greater is required. Get it with "Microsoft C++ Build Tools": https://visualstudio.microsoft.com/visual-cpp-build-tools/
>       [end of output]
> 
>   note: This error originates from a subprocess, and is likely not a problem with pip.
> error: legacy-install-failure
> 
> × Encountered error while trying to install package.
> ╰─> lxml
> 
> note: This is an issue with the package mentioned above, not pip.
> hint: See above for output from the failure.

I've no idea what is going on with my computer. I installed every requested tool and cloned it all, but it turned out like this. Anybody has any idea how to fix this?

**Comment by timcanham** [comment]
It looks like you are trying to install under Windows? We haven't really tested that environment. You would be better off installing Ubuntu under VirtualBox or something similar.

**Reply by paemaker** [reply]
Yes, I've done it and it works perfectly! Thank you so much

**Comment by LeStarch** [comment, marked as answer]
I agree with @timcanham that the normal pattern is to install in WSL or a virtual machine of some form.

Looking at your exact issue, it seems `lxml` requires "Microsoft Visual C++ 14.0", so you could install that and then try again.  F´ will not build on Windows currently, but the GDS may run.

**Reply by paemaker** [reply]
I've done it on Ubuntu and it's working now. I guess I didn't use F' in the virtual environment as well. You said that right! Thank you.
